### PR TITLE
Update vdc storage profiles

### PR DESCRIPTION
**Pull request by dataclouder**
Fixes Issue [#648](https://github.com/vmware/terraform-provider-vcd/issues/648) "adding a storage profile requires the VDC to be replaced"

Ancillary problems related to issue fix:
* The search for storage profiles was faulty: it does not work at all for name `"*"`;
* The search could return a storage profile from a Provider VDC different from the one we were using to create the VDC.

Added methods:
* `AdminVdc.AddStorageProfile`
* `AdminVdc.AddStorageProfileWait`
* `AdminVdc.RemoveStorageProfile`
* `AdminVdc.RemoveStorageProfileWait`
* `AdminVdc.SetDefaultStorageProfile`
* `AdminVdc.GetDefaultStorageProfileReference`
* `VCDClient.GetStorageProfileByHref`
* `Client.GetStorageProfileByHref`
* `VCDClient.QueryProviderVdcStorageProfileByName`
* `Client`QueryAllProviderVdcStorageProfiles`
* `Client.QueryProviderVdcStorageProfiles`

### CODE REVIEW

File: .changes/v2.13.0/393-features.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+* Added method `AdminVdc.AddStorageProfile` [GH-393]
+* Added method `AdminVdc.AddStorageProfileWait` [GH-393]
+* Added method `AdminVdc.RemoveStorageProfile` [GH-393]
+* Added method `AdminVdc.RemoveStorageProfileWait` [GH-393]
+* Added method `AdminVdc.SetDefaultStorageProfile` [GH-393]
+* Added method `AdminVdc.GetDefaultStorageProfileReference` [GH-393]
+* Added method `VCDClient.GetStorageProfileByHref` [GH-393]
+* Added method `Client.GetStorageProfileByHref` [GH-393]
+* Added method `VCDClient.QueryProviderVdcStorageProfileByName` [GH-393]
+* Added method `ClientQueryAllProviderVdcStorageProfiles` [GH-393]
```
Comment (vbauzysvmware) [inline .changes/v2.13.0/393-features.md:10]: Dot is missing 
```suggestion
* Added method `Client.QueryAllProviderVdcStorageProfiles` [GH-393]
```

File: govcd/adminvdc.go [hunk 2]
```diff
@@ -424,3 +424,171 @@ func (vdc *AdminVdc) UpdateStorageProfile(storageProfileId string, storageProfil
...
+		return Task{}, fmt.Errorf("cannot remove VDC storage profile: storage profile '%s' not found in VDC", storageProfileName)
+	}
+
+	vdcStorageProfileDetails, err := vdc.client.GetStorageProfileByHref(storageProfile.HREF)
+	if err != nil {
+		return Task{}, fmt.Errorf("cannot retrieve VDC storage profile '%s' details: %s", storageProfileName, err)
+	}
+	if vdcStorageProfileDetails.Enabled {
+		_, err = vdc.UpdateStorageProfile(extractUuid(storageProfile.HREF), &types.AdminVdcStorageProfile{
+			Name:    storageProfile.Name,
```
Comment (vbauzysvmware) [inline govcd/adminvdc.go:488]:
```suggestion
			Name:    vdcStorageProfileDetails.Name,
```

File: govcd/adminvdc.go [hunk 3]
```diff
@@ -424,3 +424,171 @@ func (vdc *AdminVdc) UpdateStorageProfile(storageProfileId string, storageProfil
...
+			Units:   "MB",
```
Comment (vbauzysvmware) [inline govcd/adminvdc.go:489]: why not using vdcStorageProfileDetails.units ?

File: govcd/adminvdc.go [hunk 4]
```diff
@@ -424,3 +424,171 @@ func (vdc *AdminVdc) UpdateStorageProfile(storageProfileId string, storageProfil
...
+	if storageProfile == nil {
+		return fmt.Errorf("cannot set VDC default storage profile: storage profile '%s' not found in VDC", storageProfileName)
+	}
+
+	vdcStorageProfileDetails, err := vdc.client.GetStorageProfileByHref(storageProfile.HREF)
+	if err != nil {
+		return fmt.Errorf("cannot retrieve VDC storage profile '%s' details: %s", storageProfileName, err)
+	}
+	_, err = vdc.UpdateStorageProfile(extractUuid(storageProfile.HREF), &types.AdminVdcStorageProfile{
+		Name:    storageProfile.Name,
```
Comment (vbauzysvmware) [inline govcd/adminvdc.go:556]: similar comments here as above

File: govcd/adminvdc.go [hunk 5]
```diff
@@ -424,3 +424,171 @@ func (vdc *AdminVdc) UpdateStorageProfile(storageProfileId string, storageProfil
...
+}
+
+// GetDefaultStorageProfileReference finds the default storage profile for the VDC
+func (adminVdc *AdminVdc) GetDefaultStorageProfileReference() (*types.Reference, error) {
+	var defaultSp *types.Reference
+	if adminVdc.AdminVdc.VdcStorageProfiles == nil || adminVdc.AdminVdc.VdcStorageProfiles.VdcStorageProfile == nil {
+		return nil, fmt.Errorf("no storage profiles found in VDC %s", adminVdc.AdminVdc.Name)
+	}
+	for _, sp := range adminVdc.AdminVdc.VdcStorageProfiles.VdcStorageProfile {
+		fullSp, err := adminVdc.client.GetStorageProfileByHref(sp.HREF)
```
Comment (vbauzysvmware) [inline govcd/adminvdc.go:579]: what here means `full` ?

File: govcd/system.go [hunk 6]
```diff
@@ -731,6 +742,47 @@ func GetStorageProfileByHref(vcdClient *VCDClient, url string) (*types.VdcStorag
...
+		"filterEncoded": "true",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Note: pageSize of 128 should be enough to get all storage profiles.
+	// In case this is not true, we trap the error, so that we become aware that this assumption is incorrect.
+	// TODO: convert this query into a cumulativeQuery
+	if results.Results.Total > 128.0 {
```
Comment (vbauzysvmware) [inline govcd/system.go:763]: Add comment here that 128 is max value which API supports. For next time we forget :)

File: govcd/system.go [hunk 7]
```diff
@@ -731,6 +742,47 @@ func GetStorageProfileByHref(vcdClient *VCDClient, url string) (*types.VdcStorag
...
+// There are four cases:
+// 1. [FOUND] The name matches and is unique among all the storage profiles
+// 2. [FOUND] The name matches, it is not unique, and it is disambiguated by the provider VDC HREF
+// 3. [NOT FOUND] The name matches, is not unique, but no Provider HREF was given: the search will fail
+// 4. [NOT FOUND] The name does not match any of the storage profiles
+func (vcdCli *VCDClient) QueryProviderVdcStorageProfileByName(name, providerVDCHref string) (*types.QueryResultProviderVdcStorageProfileRecordType, error) {
+	results, err := vcdCli.QueryWithNotEncodedParams(nil, map[string]string{
+		"type":          "providerVdcStorageProfile",
+		"pageSize":      "128",
+		"filterEncoded": "true",
```
Comment (vbauzysvmware) [inline govcd/system.go:754]: not sure we need this as we don't send any filter

**Review by vbauzysvmware — APPROVED** [review]

File: govcd/adminvdc.go (hunk 5 above)
Comment (dataclouder) [reply govcd/adminvdc.go:579]: "full", as opposed to the simple reference that we had until that point.

File: govcd/adminvdc.go (hunk 3 above)
Comment (dataclouder) [reply govcd/adminvdc.go]: changed

File: govcd/adminvdc.go (hunk 4 above)
Comment (dataclouder) [reply govcd/adminvdc.go]: changed

File: govcd/system.go (hunk 6 above)
Comment (dataclouder) [reply govcd/system.go:762]: done

File: govcd/system.go (hunk 7 above)
Comment (dataclouder) [reply govcd/system.go]: thanks. Removed

File: govcd/adminvdc.go (hunk 2 above)
Comment (dataclouder) [reply govcd/adminvdc.go]: changed

File: .changes/v2.13.0/393-features.md (hunk 1 above)
Comment (dataclouder) [reply .changes/v2.13.0/393-features.md]: fixed

File: govcd/adminvdc.go [hunk 8]
```diff
@@ -424,3 +424,171 @@ func (vdc *AdminVdc) UpdateStorageProfile(storageProfileId string, storageProfil
...
+		return err
+	}
+	return vdc.Refresh()
+}
+
+// SetDefaultStorageProfile sets a given storage profile as default
+// This operation will automatically unset the previous default storage profile.
+func (vdc *AdminVdc) SetDefaultStorageProfile(storageProfileName string) error {
+	if vdc.client.VCDHREF.String() == "" {
+		return fmt.Errorf("cannot Set VDC default storage profile: VCD HREF is unset")
```
Comment (lvirbalas) [inline govcd/adminvdc.go:538]:
```suggestion
		return fmt.Errorf("cannot set VDC default storage profile: VCD HREF is unset")
```

File: govcd/adminvdc.go [hunk 9]
```diff
@@ -424,3 +424,171 @@ func (vdc *AdminVdc) UpdateStorageProfile(storageProfileId string, storageProfil
...
+		Limit:   vdcStorageProfileDetails.Limit,
+		Default: true,
+		Enabled: takeBoolPointer(true),
+		ProviderVdcStorageProfile: &types.Reference{
+			HREF: vdcStorageProfileDetails.ProviderVdcStorageProfile.HREF,
+		},
+	},
+	)
+	if err != nil {
+		return fmt.Errorf("cannot Set VDC default storage profile '%s': %s", storageProfileName, err)
```
Comment (lvirbalas) [inline govcd/adminvdc.go:567]:
```suggestion
		return fmt.Errorf("cannot set VDC default storage profile '%s': %s", storageProfileName, err)
```

**Review by lvirbalas — APPROVED** [review]
Neat! Only polishes in review.

**Review by Didainius — APPROVED** [review]

**Pull request merged**